sqlglot/optimizer/scope.py

```diff
@@ -548,30 +565,38 @@ def _traverse_select(scope):
 
 
 def _traverse_union(scope):
-    yield from _traverse_ctes(scope)
+    prev_scope = None
```

Comment:
how come you need this prev_scope concept?

Reply:
I ended up doing it like this because if e.g. you have UNION(UNION(a, b), c) then you first visit a, b so you can construct the inner union scope, and then you visit c and can use the UNION(a, b) and c to construct the outer union scope. So basically, prev is used to keep track of the "last constructed" left scope tree. I also tried a couple approaches here but didn't turn out as well, we can chat more about this.